packages/authentication/src/services/token.service.ts

```diff
@@ -11,10 +11,32 @@ import {UserProfile} from '@loopback/security';
 export interface TokenService {
   /**
    * Verifies the validity of a token string and returns a user profile
+   *
+   * @param token The token/secret which should be validated/verified.
+   * @param userProfile Depending on the token system this optional parameter
+   *   can be used to check if the token is valid for the given user.
+   *
+   * @returns The UserProfile which belongs to the given token.
    */
-  verifyToken(token: string): Promise<UserProfile>;
+  verifyToken(token: string, userProfile?: UserProfile): Promise<UserProfile>;
```

Comment:
In what case the user profile is available for the token is verified?

Reply:
For example for `refreshToken` use case, its required to verify if the `refreshToken` if it belongs to the current user (which was verified by the `accessToken` before).

Here is an example:
https://github.com/strongloop/loopback4-example-shopping/blob/980a065224a56fd7dbd6b512519c944bd0415991/packages/shopping/src/controllers/user.controller.ts#L301-L309

Reply:
@derdeka why are you validating the jwt when you are doing a refresh? The only reason to do a refresh of the access token is if it is expired in which case then `@authenticate(jwt)` throws an error anways

Reply:
@nflaig
Thanks for your feedback. I assumed it would be a good idea to extend the `accessToken` using the `refreshToken` as long as it is still valid with the benefit of improved security.

I would like to ask you for a review of https://github.com/strongloop/loopback4-example-shopping/pull/537 to have a look at the whole refresh process.

Reply:
Refresh token itself is subject to verification. I don't expect the `refresh` to require an valid access token.

Reply:
@derdeka the current implementation would mean the user has a session timeout if he is inactive for 10mins becasue of the access token expiration which is probably not the expected behaviour.

I would create a relation between a user which can have many refresh tokens e.g. one for each device he is logged in. The refresh token value should probably not be the database Id but rather be generated in a [secure way](https://gist.github.com/joepie91/7105003c3b26e65efcea63f3db82dfba). This means if a request to the refresh token endpoint happens you just look up the refresh token and if it exists and has not expired you can issue a new access token for the user to whom the refresh token belongs to. The refresh token is similiar to traditional session tokens which are basically just a secure random and unguessable string of characters.

Reply:
Feedback applied.